### PR TITLE
Upgrade Keeper Script

### PR DESCRIPTION
**Pull request by fubuloubu**
~~depends: https://github.com/eth-brownie/brownie/pull/847~~

### CODE REVIEW

**Comment by fubuloubu** [comment]
Note to reviewers: must manually install the brownie branch listed above to run

File: scripts/keep.py [hunk 1]
```diff
@@ -38,37 +38,69 @@ def main():
         assert strategy.vault() == vault.address, "Vault mismatch! [{strategy.address}]"
```
Comment (banteg) [inline scripts/keep.py:38]:
```suggestion
        assert strategy.vault() == vault.address, f"Vault mismatch! [{strategy.address}]"
```

File: scripts/keep.py [hunk 2]
```diff
@@ -38,37 +38,69 @@ def main():
...
             debt = vault.debtOutstanding(strategy) / 10 ** vault.decimals()
             print(f"[{strategy.address}] Debt Outstanding: {debt:0.3f} {symbol}")
 
-            # TODO: Actually estimate gas
-            if strategy.tendTrigger(40000 * gas_price):
+            starting_gas_price = next(gas_strategy.get_gas_price())
+
+            try:
+                tend_gas_estimate = int(
+                    GAS_BUFFER * strategy.tend.estimate_gas({"from": bot})
```
Comment (banteg) [inline scripts/keep.py:57]: If we specify `owner=bot` when instantiating contracts and set the global gas strategy `network.gas_price(gas_strategy)`, we can omit the `{"from": bot, "gas_price": gas_strategy}` arg from all the function calls.

File: scripts/keep.py [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+from brownie.network.gas.strategies import GasNowScalingStrategy
 from decimal import Decimal
 from eth_utils import is_checksum_address
 import requests
 from time import sleep
 
 
+GAS_BUFFER = 1.2
+
+gas_strategy = GasNowScalingStrategy()
```
Comment (banteg) [inline scripts/keep.py:11]: Ideally we need a custom logic to rather cancel the tx for 21k gas on gas prices spike.

**Review by banteg — APPROVED** [review]
Looks good, although keepers can work at a loss if the gas prices spike and transaction is replaced a few times.

I added a few remarks on how to make the code a bit cleaner.

File: scripts/keep.py (hunk 2 above)
Comment (fubuloubu) [reply scripts/keep.py:57]: I think it's better as explicit

File: scripts/keep.py (hunk 3 above)
Comment (fubuloubu) [reply scripts/keep.py:11]: have to add to brownie

**Pull request merged**